src/impls.rs

```diff
@@ -1490,20 +1487,131 @@ impl<T: Config> UsingTellor<AccountIdOf<T>> for Pallet<T> {
 		Self::get_data_before(query_id, timestamp).map(|(v, t)| (v.into_inner(), t))
 	}
 
-	fn get_index_for_data_after(_query_id: QueryId, _timestamp: Timestamp) -> Option<usize> {
-		todo!()
+	fn get_index_for_data_after(query_id: QueryId, timestamp: Timestamp) -> Option<usize> {
+		let mut count = Self::get_new_value_count_by_query_id(query_id);
+		if count == 0 {
+			return None
+		}
+		count.saturating_dec();
+		let mut search = true; // perform binary search
+		let mut middle = 0;
+		let mut start = 0;
+		let mut end = count;
+		// checking boundaries to short-circuit the algorithm
+		let mut timestamp_retrieved =
+			Self::get_timestamp_by_query_id_and_index(query_id, end).unwrap_or_default();
+		if timestamp_retrieved <= timestamp {
+			return None
+		}
+		timestamp_retrieved =
+			Self::get_timestamp_by_query_id_and_index(query_id, start).unwrap_or_default();
+		if timestamp_retrieved > timestamp {
+			// candidate found, check for disputes
+			search = false;
+		}
+		// since the value is within our boundaries, do a binary search
+		while search {
```

Comment:
Could get rid of search variable and just use break statements?

Reply:
Took a look, but after implementing and then following clippy suggestions it resulted in code that was harder to read so feel it is fine to keep as is, especially as it aligns closer to source contract.